fuse_constraints/include/fuse_constraints/normal_prior_pose_3d_cost_functor.h

```diff
@@ -0,0 +1,127 @@
+
+
+namespace fuse_constraints
+{
+
+/**
+ * @brief Create a prior cost function on both the 3D position and orientation variables at once.
+ *
+ * The Ceres::NormalPrior cost function only supports a single variable. This is a convenience cost function that
+ * applies a prior constraint on both the 3D position and orientation variables at once.
+ *
+ * The cost function is of the form:
+ *
+ *             ||    [  x       -     b(0)  ] ||^2
+ *             ||    [  y       -     b(1)  ] ||
+ *             ||    [  z       -     b(2)  ] ||
+ *             ||A * [  (b(4:7) * q^-1)(1)  ] ||
+ *   cost(x) = ||    [  (b(4:7) * q^-1)(2)  ] ||
+ *             ||    [  (b(4:7) * q^-1)(3)  ] ||
+ *
+ * where, the matrix A and the vector b are fixed and (x, y, z, qw, qx, qy, qz) are the components of the position and
+ * orientation variables. Note that the covariance submatrix for the quaternion is 3x3, representing errors around each
+ * imaginary component axis. In case the user is interested in implementing a cost function of the form
+ *
+ *   cost(X) = (X - mu)^T S^{-1} (X - mu)
+ *
+ * where, mu is a vector and S is a covariance matrix, then, A = S^{-1/2}, i.e the matrix A is the square root
+ * information matrix (the inverse of the covariance).
+ */
+class NormalPriorPose3DCostFunctor
+{
+public:
+  using Matrix6d = Eigen::Matrix<double, 6, 6>;
+  using Vector7d = Eigen::Matrix<double, 7, 1>;
+
+  /**
+   * @brief Construct a cost function instance
+   *
+   * @param[in] A The residual weighting matrix, most likely the square root information matrix in order
+   *              (x, y, z, qw, qx, qy, qz)
+   * @param[in] b The 3D pose measurement or prior in order (x, y, z, qw, qx, qy, qz)
+   */
+  NormalPriorPose3DCostFunctor(const Matrix6d& A, const Vector7d& b);
+
+  /**
+   * @brief Evaluate the cost function. Used by the Ceres optimization engine.
+   */
+  template <typename T>
+  bool operator()(const T* const position, const T* const orientation, T* residual) const;
+
+private:
+  Matrix6d A_;
+  Vector7d b_;
+
+  NormalPriorOrientation3DCostFunctor orientation_functor_;
+};
+
+NormalPriorPose3DCostFunctor::NormalPriorPose3DCostFunctor(const Matrix6d& A, const Vector7d& b) :
+  A_(A),
+  b_(b),
+  orientation_functor_(Eigen::Matrix3d::Identity(), b_.tail<4>())  // Delta will not be scaled
+{
+}
+
+template <typename T>
+bool NormalPriorPose3DCostFunctor::operator()(const T* const position, const T* const orientation, T* residual) const
+{
+  // Use the 3D orientation cost functor to compute the orientation delta
+  orientation_functor_(orientation, &residual[3]);
+
+  Eigen::Map<Eigen::Matrix<T, 6, 1> > residuals_map(residual);
+  residuals_map(0) = position[0] - T(b_(0));
+  residuals_map(1) = position[1] - T(b_(1));
+  residuals_map(2) = position[2] - T(b_(2));
+
+  // Scale the residuals by the square root information matrix to account for
+  // the measurement uncertainty.
+  residuals_map.applyOnTheLeft(A_.template cast<T>());
```

Comment:
Note that the covariance is being applied to the entire residual, not just the position.